Added support for the Target Blue Pill F103C8

wassfila opened the pull request:
Notes:
* I agree to the [contributer agreement](https://github.com/ARMmbed/mbed-os/blob/master/CONTRIBUTING.md).

## Description
The STM32F103RB and STM32F10C8 are identical mcu and only differ in flash size, so the support of mbed os for one and not the other is not reasonable.
The mbed-os is currently kind of if deffing the F103C8 out of the compilation, so I parsed all of the OS and added the flag support for STM32F10C8 where needed to declare things such as OS_TASKCNT,...

## Status
**READY**


## Migrations
If this PR changes any APIs or behaviors, give a short description of what *API users* should do when this PR is merged.

NO


## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
other_pr_production | [link]()
other_pr_master | [link]()


## Todos
- [x] Tests
- [ ] Documentation


## Deploy notes
Notes regarding the deployment of this PR. These should note any
required changes in the build environment, tools, compilers, etc.


## Steps to test or reproduce
Outline the steps to test or reproduce the PR here.

0xc0170 commented:
Please sign CLA: https://developer.mbed.org/contributor_agreement/

How was this addition tested? Can run ``mbed test`` for this target?

0xc0170 commented:
One more: You might want to enable this for mbed 5 in the targets.json file (this target does not have ``release_versions`` set?

theotherjimmy commented:
Is the Blue Pill the name of the target? If so, could you add something that indicates this PR adds a target to the title?

wassfila commented:
Hi,
First thanks for being active.
- The CLA is accepted.
- The release_version is here optional as the commit was more generic just for the device, although a nice recommendation, I added a new commit for that 8fc0b9eaf3dd5cf80b4001edc8191f9ebb941356. Note that to add support of the mbed 2 and 5, I had to add support for IAR and ARM which are here just compilers supporting the same cortex M3 of course.
- The target BLUEPILL_F103C8 exist already, if you mean by adding support for a device the targets inherits the support then yes I'll change the PR name

wassfila commented:
about testing I run mbed test, I'm still new to the "mbed test" so non conclusive, I tested with blinky led and other programs serial usb and so on that I know and it is working fine, if you require any other information just let me know.

theotherjimmy commented:
@wassfila could you upload the results from ARM, IAR, and GCC_ARM compilers with the mbed test command?

wassfila commented:
@theotherjimmy, 

I have run 3 tests, as I did not had any reference, I've run it for the NUCLEO_F103RB before and after my modif, then with the BLUEPILL_F103C8 after my modif.

I'm not familiar with the tests so cannot conclude, here you have attached the complete log of the test. I have noticed that USB files are missing which are only available for the F103RB

Copy: USBHAL_STM32F103RB.h
Copy: USBHAL_STM_TARGET.h

The usb device is identical between F103C8 and F103RB devices, I even tested the USB example compiled for the F103RB device running successfully on the bluepill F103C8. So I could simply add the USB files copy and support but that is not the blocker, the block is that unfortunately I do not have the IAR environment and have no ARM compiler license. Let me know if there is any alternative or support from someone to run the tests, but that might limit my contribution.

As the target was already there in targets.json : BLUEPILL_F103C8, the support of it would not come in one step and an intermediate one would be to add support for a device STM32F103C8 without the blue pill target, but that is not really interesting for me as I would still have to have my own patch and cannot directly use the original mbed-os branch.

Note that the BLUEPILL_F103C8 is an incredibly cheap board that has a wide community, that would be nice to spread the mbed-os over it. So let me know if I can help further anyhow with its support.

Please find below the test logs.

[build_Nucleo_f2a648d.txt](https://github.com/ARMmbed/mbed-os/files/984428/build_Nucleo_f2a648d.txt)
[build_Blue_after.txt](https://github.com/ARMmbed/mbed-os/files/984427/build_Blue_after.txt)
[build_Nucleo_after.txt](https://github.com/ARMmbed/mbed-os/files/984429/build_Nucleo_after.txt)

0xc0170 commented:
@wassfila Looks like the above .txt attached are build logs, not test .CA. Can you run ``mbed test`` ?  What debugger does the board has ?

wassfila commented:
Hi,
This board is using an ST-Link-v2-1 debugger from a nucleo separate board. see the link below for more details.
I have run the test command from within the modified "mbed-os" directory as below
`D:\Projects\STM32\mbed\mbed-os>mbed test -c -m BLUEPILL_F103C8`
After the test compilation, the device is detected but I have the issue that no platform matching tests were found, could you please help me on how to launch the test correctly of where should I include the platform to the tests that have to be found 
`mbedgt: greentea test automation tool ver. 1.2.5
mbedgt: test specification file 'D:\Projects\STM32\mbed\mbed-os\BUILD\tests\BLUEPILL_F103C8\GCC_ARM\test_spec.json' (specified with --test-spec option)
mbedgt: using 'D:\Projects\STM32\mbed\mbed-os\BUILD\tests\BLUEPILL_F103C8\GCC_ARM\test_spec.json' from current directory!
mbedgt: detecting connected mbed-enabled devices...
mbedgt: detected 1 device
mbedgt: processing target 'BLUEPILL_F103C8' toolchain 'GCC_ARM' compatible platforms... (note: switch set to --parallel 1)
mbedgt: all tests finished!
mbedgt: shuffle seed: 0.8697366402
mbedgt: no platform/target matching tests were found!
mbedgt: no matching platforms were found!
mbedgt: completed in 0.03 sec
mbedgt: exited with code -110`

https://developer.mbed.org/users/hudakz/code/STM32F103C8T6_Hello/file/5ebf34d19614/mbed-STM32F103C8T6.lib

wassfila commented:
mbed test errors
`mbedgt: no platform/target matching tests were found!
mbedgt: no matching platforms were found!`
see full log for details...
[mbed-os test BLUEPILL.txt](https://github.com/ARMmbed/mbed-os/files/1002009/mbed-os.test.BLUEPILL.txt)

0xc0170 commented:
``mbed ls`` shows something ? The board should be added to ``https://github.com/NXPmicro/mbed-ls/blob/master/mbed_lstools/lstools_base.py``? Or which nucleo firmware does it use?

cc @bcostm @adustm @LMESTM @jeromecoutant Might interest you this new addition based on STM32 target.

adustm commented:
Hi,
When you connect you STLINK you are supposed to get a virtual disk mounted on it.
It should contain a file named mbed.htm. If you open this file you should reach the developper.mbed.org web page of the platform you have connected. 
Do you see this mbed.htm file ?

wassfila commented:
Hi all, and thanks all for your attention and support,

As a summary, I do not think this target is testable with "mbed test", yet I think that this change is a step forward that solves a failing compilation for a target directory that was already in the mbed-os source code.

For detailed explaination :
@0xc0170,  I assume you meant "mbedls" not "mbed ls", I installed it, but that is only to check which debug board is connected, so let me have a recapitulation :
@0xc0170, @adustm,
As described in this pull request's description, this is simply extending the support of a device for a target that's already there, I added TARGET in the title just because it was suggested by @theotherjimmy.

As you can see these changes are non functional changes for any device different that the one I added on the define, and for the one I added on the define "|| defined(TARGET_STM32F103C8)" this cannot make it worse as the mbed-os is now compiling for that device while before it was not compiling. It is now even working as I tested multiple use cases. The mbed test is not testable in this use case. It is sometimes common to fail a test not because of an issue but because of a testing issue.

To explain the situation, please have a look at the link below (mbed-STM32F103C8T6.lib) explaining the only way available today on how to use the mbed-os with the world's cheapest STM32 board ever, the BluePill. The Bluepill is an increasingly popular bread board amongst a large community of hobbyist and I selected it for my hobby projects and community of friends. The only reason why this board is popular and cheap is that it is very minimal and thus relies on other more expensive boards for debug and flashing. If mbed-os is targeting realistic applications, why would I recommend a community to pay 10 times for a debugger for every one while it is enough to have only one per person ?

To make it short, this Bluepill alone does not fulfill the requirements to be a stand alone official mbed development kit, which is to have its own embedded debugger. Here as described below, a Nucleo dev kit is beeing to flash and debug the Bluepill. It is originally a double prupose dev kit, an st-link v2-1 for its own STM32, or a stand alone st-link v2-1 by diconnecting jumpers to use with any other STM32 device.
So I do not see sense in adding the Bluepill to the "lstools_base.py" if this interface is seeing the debugger board and not the final target connected to the debugger.
So here the week spot of the mbed-os is restricting the devices support to the debuggers not to the whole supported devices by those debuggers.

I would not insist on having these changes on the os master repo, as I understand the restrictions on quality and testability, but let us just consider facts we have here:
I do not have a full solution for a clean support for the Bluepill, but that should come in incremental steps. Currently, the only recommended way is the "mbed-STM32F103C8T6.lib" which link is explained below. I find that lib solution is a dirty hack as it is overriding files from the mbed-os by precedence in the compilation includes path. That makes the Bluepill use its own headers while steeling the compilation as a NUCLEO target and having a wrong flash size. This makes even less sense if we see that in the mbed-os the TARGET_BLUEPILL_F103C8 directory is already existing under the TARGET_STM32F1 directory.

Why is the directory TARGET_BLUEPILL_F103C8 in the os if the mbed-os does not even compile for the device of that target? Putting it that way, this change is only improving the situation by making the mbed-os compile for the device STM32F103C8 which is used by that target.
I understand that the online compiler would not accept to add a dev kit that does not have its own mbed debugger, but the mbed-os being a nice cli tool, I don't see reasons for restricting it to targets which necessarily have debuggers connected, that would exclude any serious project that wants to be cost efficient, which is a majority in the IoT world. That reduce chances of the mbed-os selection, and it is also under estimating a large community of hobbyist and others who are still able to connect a board to a debugger.

I have patched the mbed-os in my own fork and have no issues merging the original from time to time, but that would be an unnessecary fragmentation of the os across communities which I would like to reduce.

https://developer.mbed.org/users/hudakz/code/STM32F103C8T6_Hello/file/5ebf34d19614/mbed-STM32F103C8T6.lib

theotherjimmy commented:
@wassfila Thanks for the detailed explanation. My understanding is that the Bluepill does not have an integrated debugger, and therefore should not be added to the mbedls database. @bridadan Could you help out with getting mbedgt to test on a target like this?

0xc0170 commented:
@wassfila Thanks for the explanation. This is similar to teensy that is already supported. The one way to test it is manual testing of the basic functionality.

Note, that this patch is based on the current rtos that is being replaced. Therefore this patch will need to be rebased one more time, once that update is integrated (currently being tested). We will let you know once that happens. Thanks for understanding.

wassfila commented:
@0xc0170  That sounds cool. Regarding manual testing of basic fuctionnalities, I'm building Hello world samples [here](https://github.com/wassfila/STM32_IoT_HelloWorld). That started using my fork with this commit for the last 5 samples.
Some samples even have log folders with [logic analyzer logs](https://github.com/wassfila/STM32_IoT_HelloWorld/tree/master/11_servo_uart_control/logs). Tests cover : Serial, DigitalOut, printf, wait, SPI with nRf lib, Servo which is using PWM, Ticker,...
Brief description [here](https://hackaday.io/project/21045-stm32-bluepill-frameworks-evaluation/log/59522-stm32-hello-world-and-ps4-joystick-to-servo)

No problem, for repatching or rebasing, just let me know.

Outside the scope of this commit but related to this Bluepill Board, the mbed cli has won [the selection](https://hackaday.io/project/21045-stm32-bluepill-frameworks-evaluation) over the platformIO framework, and I'll be creating and collecting Bluepill expansion board designs [here](https://hackaday.io/project/21396-stm32-blue-pill-iot-expansion-boards)

bridadan commented:
@wassfila Thanks for detailed explanation and patience :)

@theotherjimmy I think it is already possible to test with an external ST-Link, it just takes a little setup with mbed-ls.

You first need to mock the target: https://github.com/ARMmbed/mbed-ls#mocking-new-or-existing-target-to-custom-platform-name. This basically tells mbed-ls which MCU is attached on the other end of the generic ST-Link debugger.

If you have to hook up a separate USB serial adapter (if the ST-Link doesn't already have a serial connection), then you'll also need to "retarget" the board. This basically just tells mbed-ls what serial port should be associated with the board. The steps for that are here: https://github.com/ARMmbed/mbed-ls#retarget-mbed-ls-autodetection-results

After these two steps and mbed-ls is reporting the correct target name, mount point, and serial port, you should be good to go with `mbed test`.

wassfila commented:
@bridadan, thanks for the directives, you might also need a bit of patience as I'm new to this mbed test framework.
The target mocking is working fine, I did not need the serial port retarget as the ST-Link-v2-1 has Uart Tx-Rx pins for that purpose. 
But I had two issues, second of which I'm going to need sometime to debug.
The first is about the baud rate, I'm confused how the usb debugger handles it, but I had continuous timeouts, when I debugger in verbose mode, I picked up an "mbedhtrun" command that is passing a baud rate as argument, set to -p COM6:9600, when I change it to -p COM6:115200 then at least I see the message from the Device Under Test.
The line out of the attached logs
`[1495481961.22][CONN][RXD] Hello World!`
But then, a new preamble is sent and it gets stuck, so I have to gather more details about the test first to check if this is related to reset issue or something else. A manual reset do not help neither.
By the way I wasn't sure where to place the mbedls.json nor if it is possible to change the baudrate directly from there, if yes, what would be the parameter name then ? In the meanwhile I have another problem with the direct test with mbedhtrun.

bridadan commented:
@wassfila Good to see you picked up on the `mbedhtrun` command, that's definitely integral to the testing process. All of the mbed OS tests should be using the 9600 baud rate, how come you need to set it to 115200? Are you trying to run a different test?

Could you please share what command you're running?

Also, the test case listed in your log (the one located in `features/frameworks/utest/TESTS/unit_tests/basic_test`) doesn't print the string `Hello World!` (as you can see in its source here: https://github.com/ARMmbed/mbed-os/blob/master/features/frameworks/utest/TESTS/unit_tests/basic_test/main.cpp). Have you modified this test? Or perhaps this just means a previous binary is running, thus indicating that the device didn't actually flash properly.

wassfila commented:
@bridadan, it was a bit more subtile, the Nucleo MCU was sending that message, the debug UART had solder bridges, so now to be safe I'm using another UART port, retargetted with mbed.json, I also tested a sample on the Bluepill DUT with UART to check the PIO A2 and A3 are working fine till the PC test console at 9600. I checked the flashing, manual copy to the drive is well taking the binary into account, and when opening the drive to check I see the current test in there.
I nevertheless still do have timeouts, as seen in the logs attached, I'll have to take the time to read the docs and understand how a basic mbedgt test work to debug the test environment before I get back to running this mbed-os test.
In the meanwhile any suggestions are welcome. If it helps the issues visibility we could close this ticket until I get the test setup running, I will have to recommit after the new rtos anyway,...
[mbed_test_mock_retarget.txt](https://github.com/ARMmbed/mbed-os/files/1027386/mbed_test_mock_retarget.txt)

bridadan commented:
@wassfila sorry for my late response!

Taking a look at the log it looks there are quite a few copy errors occurring:
```
[1495663432.03][COPY][ERR] [ret=1] Command: ['copy', 'BUILD\\tests\\BLUEPILL_F103C8\\GCC_ARM\\features\\TESTS\\filesystem\\util_block_device\\util_block_device.bin', 'J:util_block_device.bin']
```

You said you were able to copy OK using drag and drop? Can you use the Windows `copy` command and see what you get there? That's how the test tools copy the file to the drive.

0xc0170 commented:
Whats the status of this PR? It should be rebased and retested with the latest cmsis5 update.

wassfila commented:
@0xc0170 I think this PR can simply be closed. Until I rebase the changes, and solve the test environment issues. At the moment I am unable to run any test due to the mentionned above copy issues, I have to solve them first.
